cloud-controller-manager/do/certificates.go

```diff
@@ -16,8 +16,199 @@ limitations under the License.
+
+		if err != nil {
+			respErr, ok := err.(*godo.ErrorResponse)
+			if !ok || respErr.Response.StatusCode != http.StatusNotFound {
+				return fmt.Errorf("failed to get DO certificate for service: %s", err)
+			}
+		}
+
+		if certificate == nil {
+			klog.V(2).Infof("service certificate is no longer valid: %s", serviceCertID)
+			serviceCertID = ""
+			updateServiceAnnotation(service, annDOCertificateID, serviceCertID)
+		}
+	}
+
+	domain, err := getDomain(service)
+	if err != nil {
+		return err
+	}
+
+	// no domain annotation, no need to ensure certificate for domain
+	if domain == nil {
+		return nil
+	}
+
+	// if the current cert represents the domain, save making an extra network request
+	// this case will arise when certificates are automatically rotated
+	if certificate != nil {
+		for _, dnsName := range certificate.DNSNames {
+			if dnsName == domain.full {
+				// we found matching certificate, break out of ensureCertificate
+				return nil
+			}
+		}
+
+		// the current certificate does not match the current domain. clear it so
+		// we can either find the matching cert or create a new one below
+		klog.V(2).Infof("service certificate is not valid for domain[%s]: %s", domain.full, serviceCertID)
+		certificate = nil
+	}
+
+	certificates, _, err := l.resources.gclient.Certificates.List(ctx, &godo.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to retrieve certificates: %s", err)
+	}
+
+findCert:
+	for _, cert := range certificates {
+		for _, dnsName := range cert.DNSNames {
+			if dnsName == domain.full {
+				klog.V(2).Infof("found existing certificate for domain[%s]: %s", domain.full, cert.ID)
+				certificate = &cert
+				break findCert
+			}
+		}
+	}
+
+	if certificate == nil {
+		certName := getCertificateName(domain.full)
+		dnsNames := []string{domain.root}
+
+		if domain.sub != "" {
+			dnsNames = append(dnsNames, domain.full)
+		}
+
+		certificateReq := &godo.CertificateRequest{
+			Name:     certName,
+			DNSNames: dnsNames,
+			Type:     certTypeLetsEncrypt,
+		}
+
+		klog.V(2).Infof("generating new service certificate for domain: %s", domain.full)
+		certificate, _, err = l.resources.gclient.Certificates.Create(ctx, certificateReq)
+		if err != nil {
+			return fmt.Errorf("failed to create certificate: %s", err)
+		}
+	}
+
+	updateServiceAnnotation(service, annDOCertificateID, certificate.ID)
+	return nil
+}
+
+// ensureDomainARecord ensures that if the service has a annDODomain annotation,
+// the domain has an A record for the full subdomain pointing to the loadbalancer
+func (l *loadBalancers) ensureDomainARecord(ctx context.Context, service *v1.Service, lb *godo.LoadBalancer) error {
+	domain, err := getDomain(service)
+	if err != nil {
+		return err
+	}
+
+	if domain == nil {
+		return nil
+	}
+
+	// the do loadbalancer service ensures the root domain associated with the
+	// certificate has an A record pointing to the LB so we do not need to ensure
```

Comment:
I'm not sure we can rely on this behaviour forever, since it's not clear from conversations with the LBaaS team whether or not this is intended behaviour. We should probably check for it anyway and add the record if it's not there, so that if the currently undocumented behaviour changes we won't have to scramble to backfill.